Feat/home page search

kvaker opened the pull request:
[U.P. (HomePage) Create Search section and Filter](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/train-a/search.md)
[#4](https://github.com/AngelinaBz/train-a/issues/4)
![image](https://github.com/user-attachments/assets/589147ce-fbd4-4413-b9cd-85442120f7dd)
![image](https://github.com/user-attachments/assets/f06febec-c530-4a15-b8bf-9e4ca0366686)
![image](https://github.com/user-attachments/assets/0fbf03a4-891f-4c15-9e0b-24f21ef967a4)
![image](https://github.com/user-attachments/assets/a1306c4c-20f7-4396-99a2-e25f0bf22bc5)
![image](https://github.com/user-attachments/assets/5c98613c-2245-401f-80f9-5354be4e73b2)

Search Form:
Implement a field for the initial city with autofill.
Implement a field for the final city with autofill.
Implement a field for selecting a date (optional: time) with a calendar.
Implement a form submission button that will be active only if all fields are valid.

Filter:
Implement the display of dates and days in the tabs form.
Make tabs clickable and update the list of available trips when selecting a date.
Implement visual highlighting of the selected tab.
Configure dynamic tab display based on search results. (use mock data until dev has finished writing API service for data requesting)
Implement adaptive tab design and search form for correct display on various devices.

Commit:
feat: add header component

Commit:
fix: lint debug

Commit:
feat: add logic header link

Commit:
fix: lint debug

Commit:
fix: lint\prettier\husky

Commit:
feat: trip.models for HomePage_SearchFilter

Commit:
feat: add home, search-form and search-results

Commit:
fix: debug lint

Commit:
refactor: search.service and search-form with API attribute, feat: add search-results

Commit:
feat: add style to search-form

Commit:
feat: add autosearch MapBox

Commit:
feat: add optional time, disabled button, swap, refact MapBox style

Commit:
fix: debug lint

kvaker committed:
fix: merge release conflict and feat environment

kvaker committed:
fix: debug test karma

AngelinaBz commented:
поменяй, пожалуйста, PR с main на release, если он готов к ревью

MaruyYak commented:
1. Мы обговорили архитектуру приложения и папка features в нее не входила, рекомендую перенести страницу home  в папку shared > pages 
2. В сервисе должен быть только сервис, используемые интерфейсы рекомендую вынести в отдельную папку models или types, и сама папка сервисы на уровне с pages \ components \ models etc. , а не внутри компонентов
3. поправь, пожалуйста, размеры, и опять же таки адаптив
![image](https://github.com/user-attachments/assets/3d29db2d-749b-4508-986b-20e446d100ac)
4. с точки зрения UI очень не аккуратно оформлены инпуты и очистка
![image](https://github.com/user-attachments/assets/1daa5c1c-3a0b-4bd6-b6e1-0c2cc0fac590)
Тоже UI часть, выровняй, пожалуйста, инпуты и лэйблы соответсвующим образом
![image](https://github.com/user-attachments/assets/78f8f781-779c-4317-8401-1900482f6ad8)
5. Не реализована проверка при выборе прошлых дат
 [The user must be able to select a date from a calendar that only allows future days.](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/train-a/search.md#:~:text=The%20user%20must%20be%20able%20to%20select%20a%20date%20from%20a%20calendar%20that%20only%20allows%20future%20days.)

kvaker committed:
refactor: change files structure

kvaker committed:
refactor: change API(MapBox/Yandex) and change style config

kvaker committed:
fix: debug karma

kvaker committed:
fix: merge conflict from release

kvaker committed:
feat: npm install ngrx

kvaker committed:
fix: debug stylelint

FYMG commented on train-a/src/app/shared/services/search.service.ts:
```diff
@@ -0,0 +1,38 @@
+    }
+    console.log('Request URL:', this.apiUrl, params.toString());
+
+    return this.http.get<SearchResult>(this.apiUrl, { params }).pipe(
```
Используй redux c effect пример #16

kvaker replied:
Почему-то твои комментарии не приходили на почту. Не видел этот комментарий. Постараюсь учесть

FYMG commented:
search это отдельная сущность, вынеси его из shared

kvaker committed:
feat: add api/station and MatAuto, MatForm and etc

feat: add api/station and MatAutoComplete, MatFormField and etc

feat: add api/station and MatAutoComplete, MatFormField and etc

feat: add api/station and MatAutoComplete, MatFormField and etc

kvaker committed:
refactor: search result and style to result search

refactor: search result and style to result search

kvaker committed:
feat: function groupbyClass, TotalSeats add modal route-dialog, date today

kvaker committed:
fix: debug karma test

AngelinaBz commented on train-a/src/app/core/components/header/header.component.html:
```diff
@@ -0,0 +1,30 @@
+
+    <div *ngIf="showLoginForm">
+      <!-- <app-login-form></app-login-form> -->
+    </div>
```
повторюсь, что это не надо, так как компоненты выводятся через router-outlet

AngelinaBz commented on train-a/src/app/shared/pages/home/search-results/search-results.component.html:
```diff
@@ -0,0 +1,104 @@
+
+          <div class="class-details" *ngIf="groupPricesByClass(route) as groupedPrices">
+            <div class="class-detail">
+              <span>1st Class</span>
+              <span>Price: ${{ groupedPrices['carriage1'] }}</span>
+              <span *ngIf="countTotalSeatsByClass(route)['carriage1'] as totalSeats">Seats: {{ totalSeats }}</span>
+            </div>
+
+            <div class="class-detail">
+              <span>2nd Class</span>
+              <span>Price: ${{ groupedPrices['carriage2'] }}</span>
+              <span *ngIf="countTotalSeatsByClass(route)['carriage2'] as totalSeats">Seats: {{ totalSeats }}</span>
+            </div>
+
+            <div class="class-detail">
+              <span>Women Only</span>
```
Ты берешь какие-то классы по умолчанию (1st, 2nd, Women Only), но их нет в ответе. Нужно показывать места и price в вагонах, которые получаем из ответа. Посмотри, пожалуйста, еще раз Response.

AngelinaBz commented on train-a/src/app/shared/pages/home/search-results/search-results.component.ts:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  groupPricesByClass(route: Route): { carriage1: number; carriage2: number; carriage3: number } {
+    const aggregatedPrices = { carriage1: 0, carriage2: 0, carriage3: 0 };
+    const count = { carriage1: 0, carriage2: 0, carriage3: 0 };
```
а если классов вагонов больше, чем три? Или в этом маршруте нет какого-то из вагонов? Опять же посмотри на Response.

AngelinaBz commented on train-a/src/app/shared/pages/home/search-results/search-results.component.ts:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  countTotalSeatsByClass(route: Route): { carriage1: number; carriage2: number; carriage3: number } {
+    const seatCounts = { carriage1: 0, carriage2: 0, carriage3: 0 };
+
+    const seatsPerCarriage = 29;
+
+    route.carriages.forEach((carriageType) => {
+      if (carriageType === 'carriage1') {
+        seatCounts.carriage1 += seatsPerCarriage;
+      } else if (carriageType === 'carriage2') {
+        seatCounts.carriage2 += seatsPerCarriage;
+      } else if (carriageType === 'carriage3') {
+        seatCounts.carriage3 += seatsPerCarriage;
+      }
+    });
```
здесь тоже фиксированное кол-во классов вагонов

kvaker replied:
Разбираюсь. Эти api/route api/carriage api/station... Крышу сносит...

AngelinaBz commented:
Route выводится неправильно. Станции повторяются 
<img width="1215" alt="Снимок экрана 2024-08-29 в 14 54 23" src="https://github.com/user-attachments/assets/20de9dfd-3cbf-4587-b70d-f0036aab34d8">

kvaker committed:
fix: debug merge conflict release

kvaker committed:
feat: add home to route

FYMG commented:
А зачем ты диалог делать начал? #6 за мной закреплена

kvaker commented:
Уже увидел. Твой пул реквест... Как начал https://github.com/rolling-scopes-school/tasks/blob/master/tasks/train-a/search.md так и продолжал. Не досмотрел. Смотрю ты разобрался с запросами и получением данных api/route api/carriage api/station

kvaker committed:
Merge branch 'release' into feat/HomePage_SearchFilter

kvaker committed:
Merge branch 'release' into feat/HomePage_SearchFilter

kvaker committed:
fix: debug lint

kvaker committed:
feat: add header link state and logo

kvaker committed:
feat: add header.comp.spec

kvaker committed:
fix: debug header.comp.spec

kvaker committed:
refactor: change modal - route-details-modal replace search-form/results in search

kvaker committed:
fix: change count seats in carriage

kvaker committed:
refactor: change count per seats and available

kvaker committed:
fix: merge conflict release

kvaker committed:
fix: debug angular.json

FYMG commented on train-a/src/app/search/components/search-results/search-results.component.html:
```diff
@@ -0,0 +1,106 @@
+          <div class="ticket-header">
+            <div class="departure">
+              <div class="time">
+                <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Вставляй иконки через mat-icons https://fonts.google.com/icons https://material.angular.io/components/icon/overview

FYMG commented on train-a/src/app/search/components/search-results/search-results.component.ts:
```diff
@@ -0,0 +1,303 @@
+    };
+  }
+
+  private getCarriageType(carriageCode: CarriageCode): CarriageItem {
```
Есть сервис для получения этого в Carriage state

FYMG commented on train-a/src/app/search/components/search-results/search-results.component.ts:
```diff
@@ -0,0 +1,303 @@
+      if (carriageItem) {
+        const freeSeats = carriageItem.leftSeats + carriageItem.rightSeats;
+
+        if (carriageCode === 'carriage1') seatCounts.carriage1 += freeSeats;
```
Переделай, у нас carriage с сервера беруться

FYMG commented on train-a/src/app/shared/components/route-dialog/route-dialog.component.html:
Файл не используется, можно удалять

FYMG commented on train-a/src/app/shared/configs/paths.ts:
```diff
@@ -1,5 +1,5 @@
 export const paths = {
-  main: '',
+  main: 'home',
```
Home Page
Browser path: /

FYMG commented on train-a/src/app/shared/models/search.models.ts:
```diff
@@ -0,0 +1,80 @@
+  toLongitude: number;
+}
+
+export interface StationInfo {
```
Это есть в одноименных сервисах, используй готовые

FYMG commented on train-a/src/app/shared/models/trip.models.ts:
```diff
@@ -0,0 +1,24 @@
+export interface Route {
+  id: number;
+  stations: Station[];
+  wagons: Wagon[];
```
Используй названия, как в апи

FYMG commented on train-a/src/app/shared/services/search.service.ts:
```diff
@@ -0,0 +1,65 @@
+
+  constructor(private http: HttpClient) {}
+
+  search(criteria: SearchCriteria): Observable<SearchResult> {
```
Используй state с effect(ами)

FYMG commented on train-a/src/app/shared/services/search.service.ts:
```diff
@@ -0,0 +1,65 @@
+    );
+  }
+
+  getStations(): Observable<StationInfo[]> {
+    return this.http.get<StationInfo[]>(this.stationsUrl).pipe(
+      catchError((err) => {
+        console.error('Stations API error:', err);
+        return throwError(err);
+      }),
+    );
+  }
+
+  getCarriages(): Observable<Carriage[]> {
+    return this.http.get<Carriage[]>(this.carriagesUrl).pipe(
+      catchError((err) => {
+        console.error('Carriages API error:', err);
+        return throwError(err);
+      }),
+    );
+  }
+
+  getRouteInfo(routeId: number): Observable<Route> {
+    return this.http.get<Route>(`${this.routeApiUrl}/${routeId}`).pipe(
+      map((route) => {
+        return route;
+      }),
+      catchError((err) => {
+        console.error('Route Info API error:', err);
+        return throwError(err);
+      }),
+    );
+  }
+}
```
Используй готовое

FYMG commented on train-a/src/assets/img/svg/logo.svg:
Используй svg из библиотеки

kvaker committed:
Merge branch 'release' into feat/HomePage_SearchFilter

kvaker committed:
fix: merge release

kvaker committed:
Merge branch 'release' into feat/HomePage_SearchFilter

kvaker committed:
fix: debug eslint

kvaker committed:
fix: karma test

kvaker committed:
fix: debug header.comp.spec

kvaker committed:
fix: debug test

fix: debug test

kvaker committed:
fix: karma test

FYMG committed:
refactor: separate getRideFromTo

FYMG committed:
feat: fully refactor this

kvaker committed:
feat: add isRouteNotFound

kvaker committed:
refactor: card with count prices and seats

kvaker committed:
Merge branch 'release' into feat/HomePage_SearchFilter

kvaker committed:
fix: debug lint

kvaker committed:
fix: debug app.config

FYMG commented:
У тебя есть нормально сделанная карточка, зачем ты данные которое должны быть в каточке перенес в result?

FYMG commented:
Не до конца понимаю, что у тебя происходит с вагонами, у тебя есть список типов вагонов в carriages там есть имя, код, бери оттуда, а не хардкодь

kvaker committed:
fix: replace from search-result to search-card

kvaker committed:
feat: add isRideInFuture and group by class seat
